ostd/src/orpc/oqueue/ringbuffer/mpmc.rs

```diff
@@ -257,27 +234,36 @@ impl<T, const STRONG_OBSERVERS: bool, const WEAK_OBSERVERS: bool>
         // Get the slot we will be writing to
         let slot = &self.get_slot(head);
 
-        if let Some(data) = self.produce_to_noone(data, head) {
-            // Wait until the generation count matches the turn. Note that we need an even turn
-            // because we are writing.
-            let mut counter = 0;
-            while self.turn(head, false) != slot.turn.load(Ordering::Acquire) {
-                counter += 1;
-                if counter % 1024 == 0 {
-                    counter = 0;
-                    // TODO(aneesh): Revisit the need for yield - this might only be needed in our
-                    // tests that don't have preemptive scheduling.
-                    Task::yield_now();
-                }
-                core::hint::spin_loop();
+        // Wait until the generation count matches the turn. Note that we need an even turn
+        // because we are writing.
+        let mut counter = 0;
+        while self.turn(head, false) != slot.turn.load(Ordering::Acquire) {
+            counter += 1;
+            if counter % 1024 == 0 {
+                counter = 0;
+                // TODO(aneesh): Revisit the need for yield - this might only be needed in our
+                // tests that don't have preemptive scheduling.
+                Task::yield_now();
             }
-            unsafe { slot.store(data) };
+            core::hint::spin_loop();
+        }
+        unsafe { slot.store(data) };
+        if self.tail.load(Ordering::Acquire) > head
+            && self
+                .strong_observer_tails
+                .iter()
+                .all(|t| t.load(Ordering::Acquire) > head)
+        {
+            // Mark the slot as writable so that the produce isn't blocked
+            slot.turn
+                .store(self.next_turn(head, false), Ordering::Release);
```

Comment:
It might be worth moving some of your comments here.

It's also not obvious to me why "> head" is appropriate here instead of  `== MAX`.